File: project2_h2o.py
#!/usr/bin/env python2

import math
import numpy as np
import scipy.linalg as spl

def print_sq_mat(mat):
    """Pretty-print a square matrix in a traditional format."""
    dim = mat.shape[0]
    for i in range(dim+1):
        print("{0:10d}".format(i)),
    print("")
    for i in range(dim):
        print("{0:10d}".format(i+1)),
        for j in range(dim):
            print("{0:10f}".format(mat[i][j])),
        print("")

"""
Step 0: Read the Coordinate Data (piratechem)
"""
import piratechem as pc
h2o = pc.molecule.Molecule("water")
pc.read.xyz("h2o_geom.xyz", h2o)
natom = h2o.num_atoms()

"""
Step 1: Read the Coordinate Data
"""

"""
Step 2: Read the Cartesian Hessian Data
"""
hessfile = open("h2o_hessian.txt", "r")
hessatom = int(hessfile.readline())
H = np.zeros((3*natom, 3*natom))
for i in range(0,3*natom):
    for j in range(0,natom):
        line = hessfile.readline().split()
        H[i][3*j] = float(line[0])
        H[i][3*j+1] = float(line[1])
        H[i][3*j+2] = float(line[2])

print("Hessian:")
print_sq_mat(H)
print("")

"""
Step 3: Mass-Weight the Hessian Matrix
"""
Hmw = np.zeros((3*natom, 3*natom))
for i in range(0,natom):
    for j in range(0,natom):
        mi = h2o.atoms[i].m
        mj = h2o.atoms[j].m
        mimj = math.sqrt(mi*mj)
        Hmw[i*natom+0][j*natom+0] = H[i*natom+0][j*natom+0]/mimj
        Hmw[i*natom+0][j*natom+1] = H[i*natom+0][j*natom+1]/mimj
        Hmw[i*natom+0][j*natom+2] = H[i*natom+0][j*natom+2]/mimj
        Hmw[i*natom+1][j*natom+0] = H[i*natom+1][j*natom+0]/mimj
        Hmw[i*natom+1][j*natom+1] = H[i*natom+1][j*natom+1]/mimj
        Hmw[i*natom+1][j*natom+2] = H[i*natom+1][j*natom+2]/mimj
        Hmw[i*natom+2][j*natom+0] = H[i*natom+2][j*natom+0]/mimj
        Hmw[i*natom+2][j*natom+1] = H[i*natom+2][j*natom+1]/mimj
        Hmw[i*natom+2][j*natom+2] = H[i*natom+2][j*natom+2]/mimj

print("Mass-weighted Hessian:")
print_sq_mat(Hmw)
print("")

"""
Step 4: Diagonalize the Mass-Weighted Hessian Matrix
"""
omega = spl.eigvalsh(Hmw)

print("Mass-weighted Hessian eigenvalues:")
for i in omega:
    print("{0:10f}".format(i))
print("")

"""
Step 5: Compute the Harmonic Vibrational Frequencies
"""
bohr2m = 0.529177249e-10

freq = np.sqrt(omega)

print("Harmonic vibrational frequencies [cm]^-1:")
for i in freq:
    print("{0:10f}".format(i))
print("")

File: project2_h2o.cpp
#include <cstdio>
#include <cmath>
#include "molecule.hpp"
#include "utils.hpp"
#include "constants.hpp"

int main()
{
  /** Step 1: Read the Coordinate Data
   *
   */
  FILE *xyzfile;
  xyzfile = fopen("h2o_geom.txt", "r");
  int natom;
  fscanf(xyzfile, "%d", &natom);
  Molecule mol(natom, 0);
  for (int i = 0; i < natom; i++)
    fscanf(xyzfile, "%lf %lf %lf %lf", &mol.zvals[i], &mol.geom[i][0], &mol.geom[i][1], &mol.geom[i][2]);
  fclose(xyzfile);

  /** Step 2: Read the Cartesian Hessian Data
   *
   */
  FILE *hessfile;
  hessfile = fopen("h2o_hessian.txt", "r");
  int hessatom;
  fscanf(hessfile, "%d", &hessatom);
  if (fabs(natom-hessatom) > 0) {
    printf("The number of atoms doesn't match.\n");
    return -1;
  }
  double **H = new double* [3*natom];
  for (int i = 0; i < 3*natom; i++)
    H[i] = new double[3*natom];
  for (int i = 0; i < 3*natom; i++)
    for (int j = 0; j < natom; j++)
      fscanf(hessfile, "%lf %lf %lf", &H[i][3*j], &H[i][3*j+1], &H[i][3*j+2]);
  fclose(hessfile);

  printf("Hessian:\n");
  print_mat(H, 3*natom, 3*natom);
  printf("\n");

  /** Step 3: Mass-Weight the Hessian Matrix
   * Divide each element of the Hessian matrix by the product of square roots of the masses of the atoms associated with the given coordinates:
   *  \vect{F}_{M}^{ij} = \frac{F_{ij}}{\sqrt{m_{i}m_{j}}}
   */
  double **Hmw = new double* [3*natom];
  for (int i = 0; i < 3*natom; i++)
    Hmw[i] = new double[3*natom];
  double mi, mj, mimj;
  for (int i = 0; i < natom; i++) {
    for (int j = 0; j < natom; j++) {
      mi = masses[(int)mol.zvals[i]]; mj = masses[(int)mol.zvals[j]];
      mimj = sqrt(mi*mj);
      Hmw[i*natom+0][j*natom+0] = H[i*natom+0][j*natom+0]/mimj;
      Hmw[i*natom+0][j*natom+1] = H[i*natom+0][j*natom+1]/mimj;
      Hmw[i*natom+0][j*natom+2] = H[i*natom+0][j*natom+2]/mimj;
      Hmw[i*natom+1][j*natom+0] = H[i*natom+1][j*natom+0]/mimj;
      Hmw[i*natom+1][j*natom+1] = H[i*natom+1][j*natom+1]/mimj;
      Hmw[i*natom+1][j*natom+2] = H[i*natom+1][j*natom+2]/mimj;
      Hmw[i*natom+2][j*natom+0] = H[i*natom+2][j*natom+0]/mimj;
      Hmw[i*natom+2][j*natom+1] = H[i*natom+2][j*natom+1]/mimj;
      Hmw[i*natom+2][j*natom+2] = H[i*natom+2][j*natom+2]/mimj;
    }
  }

  printf("Mass-weighted Hessian:\n");
  print_mat(Hmw, 3*natom, 3*natom);
  printf("\n");

  /** Step 4: Diagonalize the Mass-Weighted Hessian Matrix
   * Compute the eigenvalues of the mass-weighted Hessian:
   *  \vect{F}^{M}\vect{L} = \vect{L}\vect{\Lambda}
   */
  double *evals = new double[3*natom];
  for (int i = 0; i < 3*natom; i++) evals[i] = 0.0;
  double **evecs = new double* [3*natom];
  for (int i = 0; i < 3*natom; i++) evecs[i] = new double[3*natom];
  diag(3*natom, 3*natom, Hmw, evals, false, evecs, 1e-19);
  for (int i = 0; i < 3*natom; i++) delete[] evecs[i];
  delete[] evecs;

  printf("Mass-weighted Hessian eigenvalues:\n");
  for (int i = 0; i < 3*natom; i++)
    printf("%12.10f\n", evals[i]);
  printf("\n");

  /** Step 5: Compute the Harmonic Vibrational Frequencies
   * The vibrational frequencies are proportional to the square root of the eigenvalues of the mass-weighted Hessian:
   *  \omega_{i} = \textrm{constant} \times \sqrt{\lambda_{i}}
   */

  printf("Harmonic vibrational frequences [cm]^-1:\n");
  for (int i = 0; i < 3*natom; i++)
    printf("%10.4f\n", sqrt(evals[i])*vib_constant);
  printf("\n");

  /// Clean up after ourselves...
  for (int i = 0; i < 3*natom; i++) {
    delete[] H[i];
    delete[] Hmw[i];
  }
  delete[] H;
  delete[] Hmw;
  delete[] evals;

  return 0;
}


